File: src/main/java/wraith/fabricaeexnihilo/modules/base/EnchantmentContainer.java
package wraith.fabricaeexnihilo.modules.base;

import net.minecraft.enchantment.Enchantment;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NbtCompound;
import net.minecraft.registry.Registries;
import net.minecraft.util.Identifier;

import java.util.HashMap;
import java.util.Map;

public class EnchantmentContainer {

    private final Map<Identifier, Integer> enchantments = new HashMap<>();

    public static void addEnchantments(ItemStack itemStack, EnchantmentContainer container) {
        container.enchantments.entrySet().stream()
                .filter(entry -> entry.getValue() > 0)
                .forEach(entry -> {
                    var enchantment = Registries.ENCHANTMENT.get(entry.getKey());
                    if (enchantment != null) {
                        itemStack.addEnchantment(enchantment, entry.getValue());
                    }
                });
    }

    public Map<Identifier, Integer> getEnchantments() {
        return enchantments;
    }

    public NbtCompound writeNbt() {
        var nbt = new NbtCompound();
        enchantments.forEach((enchantment, level) -> nbt.putInt(enchantment.toString(), level));
        return nbt;
    }

    public void readNbt(NbtCompound nbt) {
        enchantments.clear();
        nbt.getKeys().forEach(key -> enchantments.put(new Identifier(key), nbt.getInt(key)));
    }

    public int getEnchantmentLevel(Enchantment enchantment) {
        var enchantmentIdentifier = Registries.ENCHANTMENT.getId(enchantment);
        if (enchantmentIdentifier == null) {
            return 0;
        }
        return getEnchantmentLevel(enchantmentIdentifier);
    }

    public int getEnchantmentLevel(Identifier enchantment) {
        return enchantments.getOrDefault(enchantment, 0);
    }

    public void setEnchantmentLevel(Enchantment enchantment, int level) {
        var enchantmentIdentifier = Registries.ENCHANTMENT.getId(enchantment);
        if (enchantmentIdentifier == null) {
            return;
        }
        setEnchantmentLevel(enchantmentIdentifier, level);
    }

    public void setEnchantmentLevel(Identifier enchantment, int level) {
        enchantments.put(enchantment, level);
    }
}

File: src/main/kotlin/exnihilofabrico/modules/base/EnchantmentContainer.kt
package exnihilofabrico.modules.base

import net.minecraft.enchantment.Enchantment
import net.minecraft.nbt.CompoundTag
import net.minecraft.util.Identifier
import net.minecraft.util.registry.Registry

data class EnchantmentContainer(val enchantments: MutableMap<Identifier, Int> = mutableMapOf()): NBTSerializable {

    override fun toTag(): CompoundTag {
        val tag = CompoundTag()
        enchantments.forEach { k,v ->
            tag.putInt(k.toString(), v)
        }
        return tag
    }

    override fun fromTag(tag: CompoundTag) {
        tag.keys.forEach {
            enchantments[Identifier(it)] = tag.getInt(it)
        }
    }

    fun getEnchantmentLevel(enchantment: Enchantment): Int {
        return getEnchantmentLevel(Registry.ENCHANTMENT.getId(enchantment) ?: return 0)
    }

    fun getEnchantmentLevel(enchantment: Identifier): Int {
        return enchantments[enchantment] ?: 0
    }

    fun setEnchantmentLevel(enchantment: Enchantment, level: Int) {
        setEnchantmentLevel(Registry.ENCHANTMENT.getId(enchantment) ?: return, level)
    }

    fun setEnchantmentLevel(enchantment: Identifier, level: Int) {
        enchantments[enchantment] = level
    }

    fun setAllEnchantments(other: EnchantmentContainer) {
        enchantments.clear()
        other.enchantments.forEach { enchantment, level ->
            enchantments[enchantment] = level
        }
    }

}